go/ngql/types.go

```diff
@@ -13,18 +13,55 @@ import (
 
 	"github.com/attic-labs/graphql"
 	"github.com/attic-labs/noms/go/d"
-	"github.com/attic-labs/noms/go/hash"
 	"github.com/attic-labs/noms/go/types"
 )
 
+// TypeConverter provides functions to convert between Noms types and GraphQL
+// types.
+type TypeConverter struct {
+	tm       *TypeMap
+	NameFunc NameFunc
+}
+
+// NewTypeConverter creates a new TypeConverter.
+func NewTypeConverter() *TypeConverter {
+	return &TypeConverter{
+		NewTypeMap(),
+		DefaultNameFunc,
+	}
+}
+
+// NameFunc defines how to compute the GraphQL name for a Noms type.
+type NameFunc func(nomsType *types.Type, isInputType bool) string
+
+func (tc *TypeConverter) getTypeName(nomsType *types.Type) string {
+	return tc.NameFunc(nomsType, false)
+}
+
+func (tc *TypeConverter) getInputTypeName(nomsType *types.Type) string {
+	return tc.NameFunc(nomsType, true)
+}
+
+// NomsTypeToGraphQLType creates a GraphQL type from a Noms type that knows how
+// to resolve the Noms values.
+func (tc *TypeConverter) NomsTypeToGraphQLType(nomsType *types.Type) graphql.Type {
+	return tc.nomsTypeToGraphQLType(nomsType, false)
+}
+
+// NomsTypeToGraphQLInputType creates a GraphQL input type from a Noms type.
+// Input types may not be unions or cyclic structs. If we encounter those
+// this returns an error.
+func (tc *TypeConverter) NomsTypeToGraphQLInputType(nomsType *types.Type) (graphql.Input, error) {
+	return tc.nomsTypeToGraphQLInputType(nomsType)
+}
+
 // TypeMap is used as a cache in NomsTypeToGraphQLType and
 // NomsTypeToGraphQLInputType.
 type TypeMap map[typeMapKey]graphql.Type
 
 type typeMapKey struct {
-	h             hash.Hash
+	name          string // name must be unique anyway...
```

Comment:
It used to have the hash there to ensure uniqueness but the name has to unique anyway. I removed the comment.